### PR TITLE
Makman2/lintertuts

### PR DESCRIPTION
**Pull request by Makman2**
Alright my little improvement for the linter tutorial. I think there needs to be quite more written later, but to not overdrive the size of the tutorial, we should move them into different files, denoting them "Tutorial Part 2" etc.

### CODE REVIEW

**Comment by SanketDG** [comment]
ack 6ff4e20

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 1]
```diff
@@ -14,289 +14,286 @@ need to implement it on your own. Don't worry, it's easy!
...
 ::
 
-    (line):(column): (severity): (message)
+    No config file found, using default configuration
+    ************* Module coalib.bearlib.abstractions.Linter
+    L1C0: C0111 - Missing module docstring
+    L42C48: E1101 - Class 'Enum' has no 'reverse' member
+    L77C32: E1101 - Class 'Enum' has no 'reverse' member
+    L21C0: R0912 - Too many branches (16/12)
+    L121C28: W0613 - Unused argument 'filename'
```
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: would be good to have `W0613` as origin so you can deactivate single rules via ignore statements

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 2]
```diff
@@ -14,289 +14,286 @@ need to implement it on your own. Don't worry, it's easy!
...
+This class is already fully functional and allows to parse issues yielded by
+Pylint!
 
-    class HTMLTutorialBear(LocalBear, Lint):
-        executable = 'html_lint.py'
-        output_regex = re.compile(
-            r'(?P<line>\d+):(?P<column>\d+):\s'
-            r'(?P<severity>Error|Warning|Info):\s(?P<message>.+)'
-        )
+Using severities
```
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: severities -> Severities

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 3]
```diff
@@ -14,289 +14,286 @@ need to implement it on your own. Don't worry, it's easy!
...
-    our current dir: ``HTMLTutorialBear.py`` and our ``sample.html``.
+    our current dir: ``PylintTutorialBear.py`` and our ``sample.py``.
     Naming is **very** important in *coala*. *coala* will look for bears
     by their **filename** and display them based on their
     **classname**.
 
-Adding Settings to our Bear
----------------------------
+Using the ``corrected`` output format
+-------------------------------------
```
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: read about http://www.grammar-monster.com/lessons/capital_letters_title_case.htm

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 4]
```diff
@@ -14,289 +14,286 @@ need to implement it on your own. Don't worry, it's easy!
...
+                          'E': RESULT_SEVERITY.MAJOR,
+                          'C': RESULT_SEVERITY.NORMAL,
+                          'R': RESULT_SEVERITY.NORMAL,
+                          'I': RESULT_SEVERITY.INFO})
+    class PylintTutorialBear:
+        @staticmethod
+        def create_arguments(filename, file, config_file,
+                             pylint_rcfile: str=os.devnull):
+            return ('--msg_template="L{line}C{column}: {msg_id} - {msg}"',
+                    '--reports=n', '--rcfile=' + pylint_rcfile, filename)
```
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: docs missing

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 5]
```diff
@@ -305,25 +302,29 @@ By running
...
 You now know how to write a linter Bear and also how to use it in your
 project.
+
 Congratulations!
+
+Where to find more...
+---------------------
+
+If you need more information about the ``@linter`` decorator, refer to the API
+documentation. It's more versatile than described in this little tutorial.
```
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: you promised more info about the config file generation and instead you made an example with the rc file which is what we *dont* want people to do

**Comment by sils1297** [comment]
https://github.com/coala-analyzer/coala/pull/2127/commits/c0fdee674572fa402403c09817d1dfde2400f0dc needs wokr

**Comment by sils1297** [comment]
https://github.com/coala-analyzer/coala/pull/2127/commits/c0fdee674572fa402403c09817d1dfde2400f0dc needs work

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 6]
```diff
@@ -1,8 +1,8 @@
 Linter Bears
 ============
 
-Welcome. This tutorial aims to show you how to use the Lint
-class in order to integrate linters in your bears.
+Welcome. This tutorial aims to show you how to use the Linter decorator in
```
Comment (SanketDG) [inline docs/Users/Tutorials/Linter_Bears.rst]: linter decorator?

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 7]
```diff
@@ -14,289 +14,286 @@ need to implement it on your own. Don't worry, it's easy!
...
+preferred. We will do this soon for ``PylintTutorialBear``.
 
-    <html>
-      <body>
-        <h1>Hello, world!</h1>
-      </body>
-    </html>
+So which are the exact command line arguments we need to provide? It depends on
+the output format of the linter. The linter decorator is capable of handling
+different output formats:
```
Comment (SanketDG) [inline docs/Users/Tutorials/Linter_Bears.rst]: I would use `two` instead of `different`

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 8]
```diff
@@ -14,289 +14,286 @@ need to implement it on your own. Don't worry, it's easy!
...
 
-Now we can run
+- ``regex``: This parses issue messages yielded by the underlying executable.
+- ``corrected``: Auto-generates results from a fixed/corrected file provided by
+  the tool.
 
-.. Start ignoring LineLengthBear
+In this tutorial we are going to use the ``regex`` output format. But before we
+continue with modifying our bear, we need to figure out how exactly output from
+Pylint looks like so we can parse it accordingly.
```
Comment (SanketDG) [inline docs/Users/Tutorials/Linter_Bears.rst]: how output from Pylint exactly looks like

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 9]
```diff
@@ -14,289 +14,286 @@ need to implement it on your own. Don't worry, it's easy!
...
 ::
 
-    $ html_lint.py -h
+    @linter(...
+            output_format='corrected')
 
-We can see that there is a ``--disable`` option which lets us disable some
-checks. Let's add that functionality to our bear.
+If your underlying tool generates a corrected file, the class generates
+automatically patches for the changes made and yields results accordingly.
```
Comment (SanketDG) [inline docs/Users/Tutorials/Linter_Bears.rst]: automatically generates

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 5 above)
Comment (SanketDG) [inline docs/Users/Tutorials/Linter_Bears.rst]: Also, if we could link to the API documentation directly here..

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 1 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: this is a sample tutorial, you think we should use origin here and complicate things? We should do that inside the real PylintBear though^^

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 5 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: @sils1297 did I promise? :O
but right using the rc-file then is not the best then...
@SanketDG is that possible? would do that indeed^^

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 1 above)
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: yes! A tutorial shows people how things are done well. If we give them a bad example we get lots of bad bears.

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 10]
```diff
@@ -14,289 +14,286 @@ need to implement it on your own. Don't worry, it's easy!
...
 
 ::
 
-    self.arguments = '--disable=' + ignore
-    self.arguments += ' {filename}'
-    return self.lint(filename)
+    @linter(...)
+    class PylintTutorialBear:
+        """
+        Lints your Python files!
```
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: Make a good comment here, this is like setting a standard for documenting bears.

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 6 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: maybe I should use everywhere `@linter` as a term^^

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 7 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: we may implement later more formats, that's why I wanted to be general^^ Shall I change anyway?

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 8 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: see a few lines below :)

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 9 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: :+1:

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 7 above)
Comment (SanketDG) [inline docs/Users/Tutorials/Linter_Bears.rst]: right `different` is okay then

**Comment by sils1297** [comment]
can you rebase and squash?

**Comment by Makman2** [comment]
I still haven't addressed all concerns, just wanted to upload my recent stuff not to lose it :)

**Comment by sils1297** [comment]
ack (reack?) https://github.com/coala-analyzer/coala/pull/2127/commits/0f9530bf58347241ef2685b06ccb3284ef532e73

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 10 above)
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: I'd remove this line

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 11]
```diff
@@ -14,289 +14,324 @@ need to implement it on your own. Don't worry, it's easy!
...
-    self.arguments = '--disable=' + ignore
-    return self.lint(filename)
+    @linter(...)
+    class PylintTutorialBear:
+        """
+        Lints your Python files!
 
-Right place for '{filename}'
-----------------------------
+        Checks for codings standards (like well-formed variable names), detects
```
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: Checks -> Check

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 12]
```diff
@@ -14,289 +14,324 @@ need to implement it on your own. Don't worry, it's easy!
...
-Depending on where the executable (``html_lint.py`` in this case) wants the
-file-name (eg. ``sample.html``) to be present in the command which does the
-linting, we add ``'{filename}'`` to the arguments. When we run
-``html_lint.py -h``, we can see that the command signature is:
-``html5_lint.py [--disable=DISABLE] FILENAME...``
+        See http://pylint-messages.wikidot.com/all-messages for a list of all
+        checks and error codes.
 
-So, we want ``'{filename}'`` at the end of the arguments.
+        https://pylint.org/
```
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: remove this line, you refer to pylint before

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 13]
```diff
@@ -14,289 +14,324 @@ need to implement it on your own. Don't worry, it's easy!
...
+        """
```
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: while you're speaking of documentation you should also add the LANGUAGES attribute

File: docs/Users/Tutorials/Linter_Bears.rst [hunk 14]
```diff
@@ -14,289 +14,324 @@ need to implement it on your own. Don't worry, it's easy!
...
-``typed_list(item_type)`` is a function that converts the given input
-(which the user will pass to the Bear as a setting) into a list of
-items and afterwards will apply a conversion to type ``item_type`` to each
-item in the list (you can also use basic types like ``int``, ``bool``, etc.
-see :doc:`Writing Bears <Writing_Bears>`)
-Next, we have to add our setting as a parameter for the ``run()`` method
-of our bear.
-We will give the param a sugestive name like ``htmllint_ignore``.
+We can see that there is a ``--rcfile`` option which lets us specify a
+configuration file for Pylint. Let's add that functionality to our bear.
```
Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: can you at least make a note that for a good linter integration you would also use the generate config stuff? And a link to the docs for it?

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 11 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: (The bear) checks for ...
I think that's okay :O

Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: we had some discussion about this, if you want to change it trigger a new one. We agreed on imperative tense for doc comments and will soon have a bear for it.

Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: I guess/hope

File: docs/Users/Tutorials/Linter_Bears.rst (hunk 14 above)
Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: ah yeah right we wanted to change this parameter so we can use rcfile later for config stuff (though pylint supports using plain command line)...
For config stuff there's no link^^

Comment (sils1297) [inline docs/Users/Tutorials/Linter_Bears.rst]: there is API documentation

Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: when thinking about it, as pylint allows to supply everything via command line, using `create_arguments` only is preferred imo.
Maybe we should do another tutorial for this topic? But then we need to do it for another tool and the user needs to reread stuff again...

Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: @sils1297 can we open an issue for documenting `generate_config`?

Comment (Makman2) [inline docs/Users/Tutorials/Linter_Bears.rst]: @sils1297 can we open an issue for documenting `generate_config`?

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2127/commits/2e0fd83a5b7437fdc8f92a14a290e282dafa5e8e

**Comment by sils1297** [comment]
@rultor merge

**Pull request merged**